modules/compute/gke-node-pool/disk_definitions.tf

```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+## Required variables:
+#  local_ssd_count_ephemeral_storage
+#  local_ssd_count_nvme_block
+#  machine_type
+
+locals {
+
+  local_ssd_machines = {
+    "a3-highgpu-8g" = { local_ssd_count_ephemeral_storage = 16, local_ssd_count_nvme_block = null },
+    "a3-megagpu-8g" = { local_ssd_count_ephemeral_storage = 16, local_ssd_count_nvme_block = null },
+  }
+
+  generated_local_ssd_config = lookup(local.local_ssd_machines, var.machine_type, { local_ssd_count_ephemeral_storage = null, local_ssd_count_nvme_block = null })
+
+  # Select in priority order:
+  # (1) var.local_ssd_count_ephemeral_storage and var.local_ssd_count_nvme_block if any is not null
+  # (2) local.local_ssd_machines if not empty
+  # (3) default to null value for both local_ssd_count_ephemeral_storage and local_ssd_count_nvme_block
```

Comment:
Should we perhaps default to 0 instead of null and always include the blocks in the main terraform.

My thinking is that if it really falls all the way through and we use null, then we are exposed to destroying node pools on re-creation. 

I would rather just set to zero and risk an error on creation due to bad config.

The first class of problem is a hidden bug that could do damage to a running cluster later (high cost). The second class of problem will show on deployment and can be worked around with no code change by adding the value to the blueprint.

Reply:
1. Given `local_ssd_count_ephemeral_storage = 16` and `local_ssd_count_nvme_block = 0` and `including both blocks` throws error:
`Error: error creating NodePool: googleapi: Error 400: At most one Local SSD configurations can be specified.`

2. In my tests, the only acceptable scenarios for a3 machines were as follows:

- Set no blocks for ephemeral and nvme: creates node-pool successfully but re-deploy forces recreation
- Set only one block for eithe ephermal or nvme with acceptable values for the machine type: creates node-pool successfully and does not force recreation on re-deploy

3. As I learned from you, we cannot consider 0 as block exclusion because we may ignore explicit user actions.


Therefore, considering 0 as default and adding both blocks doesn't seem to be practically possible here. Please let me know if I'm missing something here.

Reply:
I'm resolving this as we agreed on the current approach over the chat.